### DISCUSSION TITLE
Detached/disrupted step matching is slow

### DISCUSSION THREAD

**Discussion by astroJeff** [POSYDON-code/POSYDON#824] [Ideas]
On average these steps each take ~0.6s per binary, and are run often. The principal reason is due to matching to the single star grids. Can we improve the algorithm to speed it up? This is the most time-intensive step in our binary population synthesis.

**Comment by astroJeff** [comment]
This is a long-term goal of ours, but becoming more and more important

**Comment by maxbriel** [comment]
I've transferred this to a discussion to make this easier to track suggestions.

A few things I've noticed in the matching that we can improve:

1. I've noticed that the first binary runs significantly slower (~5s) compared to repeat runs of a binary. Some thoughts on why this might happen: 
     a.  In `get_root0` we are building the scaler "lazily" during the first run, see [track_match.py](https://github.com/POSYDON-code/POSYDON/blob/e38d6e8201b76219fbee3e4981633ac759168aad/posydon/binary_evol/DT/track_match.py#L608-L627)
    b. We are building part of the `self.rootm` matrix as we go (in [track_match.py](https://github.com/POSYDON-code/POSYDON/blob/e38d6e8201b76219fbee3e4981633ac759168aad/posydon/binary_evol/DT/track_match.py#L477-L486)), which adds overhead on the first run as well.

Maybe there is a way to implement scipy's `(c)KDtree` somehow for the matching or finding the nearest neighbour? This might be faster than our manual setup? Not sure if this would work with our current setup.

@sgossage you've been working on this code a lot recently with the RAM usage, do you have suggestions for possible speed-ups?
